File: packages/android_intent_plus/lib/put_classes/put_parcelable_array.dart
import 'package:android_intent_plus/parcelable_classes/base/parcelable_base.dart';
import 'package:android_intent_plus/put_classes/base/put_base.dart';
import 'package:json_annotation/json_annotation.dart';

part 'put_parcelable_array.g.dart';

@JsonSerializable()
class PutParcelableArray extends PutBase {
  factory PutParcelableArray.create({
    required String key,
    required List<ParcelableBase> values,
  }) =>
      PutParcelableArray(
        key: key,
        javaClass: 'PutParcelableArray',
        values: values,
      );

  PutParcelableArray({
    required String key,
    required String javaClass,
    required this.values,
  }) : super(
          key: key,
          javaClass: javaClass,
        );

  final List<ParcelableBase> values;

  factory PutParcelableArray.fromJson(Map<String, dynamic> json) =>
      _$PutParcelableArrayFromJson(json);

  @override
  Map<String, dynamic> toJson() => _$PutParcelableArrayToJson(this);
}

File: packages/android_intent_plus/android/src/main/java/dev/fluttercommunity/plus/androidintent/Bundle/PutClasses/PutParcelableArray.java
package dev.fluttercommunity.plus.androidintent.Bundle.PutClasses;

import android.os.Parcelable;
import dev.fluttercommunity.plus.androidintent.Bundle.Constants;
import dev.fluttercommunity.plus.androidintent.Bundle.ParcelableClasses.base.ParcelableBase;
import dev.fluttercommunity.plus.androidintent.Bundle.PutClasses.base.PutBase;
import java.util.ArrayList;
import java.util.List;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class PutParcelableArray extends PutBase<List<ParcelableBase>> {

  public static final String JAVA_CLASS = "PutParcelableArray";

  public PutParcelableArray(String key, List<ParcelableBase> value) {
    super(key, JAVA_CLASS, value);
  }

  public static void convert(android.os.Bundle bundle, PutParcelableArray putParcelableArray) {
    bundle.putParcelableArray(
        putParcelableArray.key,
        PutParcelableArrayList.convertToListOfParcelable(putParcelableArray.value)
            .toArray(new Parcelable[0]));
  }

  public static PutParcelableArray fromJson(JSONObject jsonObject) throws JSONException {
    final ArrayList<ParcelableBase> arrayList = new ArrayList<>();
    final JSONArray jsonArray = jsonObject.getJSONArray(Constants.VALUE);
    for (int i = 0; i < jsonArray.length(); i++) {
      arrayList.add(PutParcelable.parcelableBaseFromJson(jsonArray.getJSONObject(i)));
    }
    return new PutParcelableArray(jsonObject.getString(Constants.KEY), arrayList);
  }

  @Override
  public JSONObject toJson() throws JSONException {
    JSONObject jsonObject = new JSONObject();
    jsonObject.put("key", key);
    jsonObject.put("javaClass", JAVA_CLASS);
    JSONArray jsonArray = new JSONArray();
    for (ParcelableBase item : value) {
      jsonArray.put(ParcelableBase.toJson(item));
    }
    jsonObject.put("value", jsonArray);
    return jsonObject;
  }
}
